subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,253 @@
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_ctrl.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/sys/base64.h>
+#include <date_time.h>
+#include "nrf_cloud_fsm.h"
+#include "nrf_cloud_mem.h"
+#include "nrf_cloud_codec_internal.h"
+#include "nrf_cloud_transport.h"
+#include "nrf_cloud_logs_internal.h"
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_logs.h>
+
+#define REST_WAIT_MS 1000
+
+LOG_MODULE_REGISTER(nrf_cloud_logs, CONFIG_NRF_CLOUD_LOGGER_LOG_LEVEL);
+
+static bool enabled;
+static int nrf_cloud_log_level = CONFIG_NRF_CLOUD_LOG_BACKEND_LEVEL;
+
+#if defined(CONFIG_NRF_CLOUD_MQTT) || defined(CONFIG_NRF_CLOUD_REST)
+static int64_t starting_timestamp;
+static atomic_t log_sequence;
+#endif
+
+static K_SEM_DEFINE(ncl_active, 1, 1);
+
+void nrf_cloud_log_init(void)
+{
+	static bool initialized;
+	int err = 0;
+
+	if (initialized) {
+		return;
+	}
+
+	(void)nrf_cloud_codec_init(NULL);
+
+	if (IS_ENABLED(CONFIG_DATE_TIME)) {
+		err = date_time_now(&starting_timestamp);
+	}
+	if (!err) {
+		initialized = true;
+	}
+}
+
+#if (defined(CONFIG_NRF_CLOUD_MQTT) || defined(CONFIG_NRF_CLOUD_REST))
+void logs_init_context(void *rest_ctx, const char *dev_id, int log_level,
+		       uint32_t src_id, const char *src_name,
+		       uint8_t dom_id, int64_t ts,
+		       struct nrf_cloud_log_context *context)
+{
+	context->rest_ctx = rest_ctx;
+	context->level = log_level;
+	context->src_id = src_id;
+	context->src_name = src_name;
+	context->dom_id = dom_id;
+	context->sequence = log_sequence++;
+
+	if (dev_id) {
+		strncpy(context->device_id, dev_id, NRF_CLOUD_CLIENT_ID_MAX_LEN);
+	} else {
+		context->device_id[0] = '\0';
+	}
+
+	if (IS_ENABLED(CONFIG_DATE_TIME)) {
+		context->ts = ts + starting_timestamp;
+	} else {
+		context->ts = 0;
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_LOG_BACKEND)
+static void log_inject(int log_level, const char *fmt, va_list ap)
+{
+	__ASSERT_NO_MSG(fmt != NULL);
+
+	/* Idempotent so is OK to call every time. */
+	nrf_cloud_log_init();
+
+	/* Cloud logging is enabled, so send it through the main logging system. */
+	z_log_msg_runtime_vcreate(Z_LOG_LOCAL_DOMAIN_ID, NULL, log_level,
+				  NULL, 0, Z_LOG_MSG2_CBPRINTF_FLAGS(0), fmt, ap);
+}
+
+#else /* !CONFIG_NRF_CLOUD_LOG_BACKEND */
+static int log_format(struct nrf_cloud_log_context *context,
+		      char *buf, struct nrf_cloud_tx_data *output,
+		      void *ctx, const char *dev_id, int log_level, const char *fmt, va_list ap)
+{
+	__ASSERT_NO_MSG(context != NULL);
+	__ASSERT_NO_MSG(buf != NULL);
+	__ASSERT_NO_MSG(output != NULL);
+	__ASSERT_NO_MSG(fmt != NULL);
+	int err;
+
+	/* Idempotent so is OK to call every time. */
+	nrf_cloud_log_init();
+
+	logs_init_context(ctx, dev_id, log_level, 0, "nrf_cloud_logs",
+			  Z_LOG_LOCAL_DOMAIN_ID, k_uptime_get(), context);
+
+	vsnprintk(buf, CONFIG_NRF_CLOUD_LOG_BUF_SIZE - 1, fmt, ap);
```

Comment:
Check whether you should rather use `vsnprintfcb` or document that without `CONFIG_NRF_CLOUD_LOG_BACKEND` printk is used which means reduced set of specifiers compared to logger that uses [Zephyr's Formatted Output](https://docs.zephyrproject.org/latest/services/formatted_output.html) framework.

We had a nasty bug in modem_shell related to this when printing with %lld. read comments above this line:
https://github.com/nrfconnect/sdk-nrf/blob/main/samples/nrf9160/modem_shell/src/utils/mosh_print.c#L106
Commit where the fix was made: https://github.com/nrfconnect/sdk-nrf/commit/950f72df64ddf668055d0c09e62f314b5478d4c4